### PR TITLE
[Dashboard] Remove gpustats dependencies from Ray[default]

### PR DESCRIPTION
**Pull request by jonathan-anyscale**
## Why are these changes needed?
Add method to get gpu utilization similarly on how gpustats did, and remove gpustats from ray[default] dependencies.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: dashboard/client/src/pages/node/index.tsx
Comment (jjyao) [inline dashboard/client/src/pages/node/index.tsx:89]: Since we vendor in so this will never happen?

File: dashboard/modules/reporter/reporter_agent.py
Comment (jjyao) [inline dashboard/modules/reporter/reporter_agent.py:400]: pynvml can never be None?

File: dashboard/modules/reporter/reporter_agent.py
Comment (jjyao) [inline dashboard/modules/reporter/reporter_agent.py:432]: This is how gpustat is implemented?

**Review by jjyao — COMMENTED** [review]
lg. Could you update the PR description?

File: dashboard/modules/reporter/reporter_agent.py
Comment (jonathan-anyscale) [reply dashboard/modules/reporter/reporter_agent.py:400]: right!

File: dashboard/modules/reporter/reporter_agent.py
Comment (jonathan-anyscale) [reply dashboard/modules/reporter/reporter_agent.py:432]: yes, calling the same pynvml methods from here
https://github.com/wookayin/gpustat/blob/64d22a9e20f6c76a85a858c87f6d7313300a6d65/gpustat/core.py#L454

**Review by jjyao — APPROVED** [review]

File: dashboard/client/src/type/node.d.ts
Comment (jjyao) [inline dashboard/client/src/type/node.d.ts:70]: so this is MB?

File: dashboard/modules/reporter/reporter_agent.py
Comment (wookayin) [inline dashboard/modules/reporter/reporter_agent.py:261]: `Megabytes` since you already have the type alias

File: dashboard/modules/reporter/reporter_agent.py
Comment (wookayin) [inline dashboard/modules/reporter/reporter_agent.py:449]: So when any pynvml error happens, this method `_get_gpu_usage` will return an empty list `[]`. Ideally throwing exceptions would be a better design, but I don't think it's a good behavior to return an empty list. You may want to return `None` instead, to be exception-safe. Depends on how other `_get_xxx` methods are designed in https://github.com/ray-project/ray/blob/master/dashboard/modules/reporter/reporter_agent.py.

File: src/ray/raylet/scheduling/cluster_resource_scheduler.cc
Comment (wookayin) [inline src/ray/raylet/scheduling/cluster_resource_scheduler.cc:49]: nit: Seems irrelevant change, just auto-formatter artifact; you may want to revert this change from this PR.

**Comment by jjyao** [comment]
@ericl needs your approval here as code owner.

Context: we are removing GPUtil and gpustats dependencies with a vendored in pyvnml single file library (the offical nvidia library that other libraries use internally) so that it works with minimal ray and it also removes unnecessary transitive dependencies included by gpustats for terminal display.

File: dashboard/modules/reporter/reporter_agent.py
Comment (wookayin) [reply dashboard/modules/reporter/reporter_agent.py:449]: Yes, that seems fair. IMHO null and empty collection should be distinguished well, but if that's the design and is consistent with other reporter fields, it should be OK. Sounds good to me!

**Review by wookayin — APPROVED** [review]

File: dashboard/modules/reporter/reporter_agent.py
Comment (jjyao) [inline dashboard/modules/reporter/reporter_agent.py:418]: @wookayin actually we need to do the same thing as https://github.com/wookayin/gpustat/pull/143/ if we want to support older version of nvidia drivers?

Comment (wookayin) [reply dashboard/modules/reporter/reporter_agent.py:418]: Yes, your understanding is correct. The current bundled version [12.535.133](https://github.com/wookayin/nvidia-ml-py/blob/12.535.133/pynvml.py (actually since 11.515.0) uses `nvmlDeviceGetGraphicsRunningProcesses_v3` API, which will raise "function not found error" for drivers < 510.39.

Similarily, the v2 API was introduced in nvidia-ml-py=11.450.129, requires NVIDIA 450.66+. Legacy drivers as low as R430 (which is pretty outdated at this point) will a pynvml version compatible with `nvidia-ml-py<11.450.129`.

So in order to support "old" NVIDIA drivers you'll need to do monkey-patch similar in wookayin/gpustat#143, or modify the `pynvml.py` bundled in ray in such a way that the function uses the v2 C function (or even the v1 function). If this is to be done without a third-party library such as gpustat or nvitop, it'd be quite bothersome.

Comment (wookayin) [reply dashboard/modules/reporter/reporter_agent.py:418]: FYI `gpustat` requires NVIDIA 450.66 (August 2020) or higher (old versions of gpustat can work with R430 drivers though) due to `nvmlDeviceGetGraphicsRunningProcesses_v2` (v2) API.

BTW, MIG support is available only in 11.450.129+, so without a very cumbersome fallback mechanism which I don't think it's worth doing, the minimum driver version ray would want to require would be 450.66 at least, which I think is a reasonable requirement.

Comment (jjyao) [reply dashboard/modules/reporter/reporter_agent.py:418]: So in order to support 450.66, we need to implement the fallback from v3 to v2. Is my understanding correct? What do you mean by `which I don't think it's worth doing`?

Comment (wookayin) [reply dashboard/modules/reporter/reporter_agent.py:418]: Exactly. I meant it is not worth the effort to support drivers even lower than 450.66 (that does not have v2 process API). We'll need some compatibility layer (either fallback as in wookayin/gpustat#143 or manually modifying `pynvml.py`).

Comment (jjyao) [reply dashboard/modules/reporter/reporter_agent.py:418]: @wookayin I guess another option is to downgrade the bundled version to one version below `11.515.0` that still users v2 API and also still has MIG support. I assume this will then work with all drivers 450.66+?

Comment (wookayin) [reply dashboard/modules/reporter/reporter_agent.py:418]: Yes. I think that would be the easiest solution but make sure you test with the actual driver instance. Something like 11.495.46? Actually, I find that the v3 API was introduced in 11.510.69 not 11.515.0 (sorry my mistake).

You will find useful:
- https://github.com/wookayin/nvidia-ml-py/blame/main/pynvml.py
- https://github.com/wookayin/nvidia-ml-py/commits/main

Comment (jonathan-anyscale) [reply dashboard/modules/reporter/reporter_agent.py:418]: @wookayin do you know the main difference between v2 and v3 APIs? seems like for our use case, v2 will be enough, but is there other component we need to aware from v3?

Comment (wookayin) [reply dashboard/modules/reporter/reporter_agent.py:418]: I think the main difference of the v3 API is the data structure on process detail. NVIDIA never properly documented this anywhere. But, I guess using v2 API should be fine for you, but I haven't really tested calling v2 API on recent NVIDIA drivers.

FYI there was [a huge mess in 535 driver versions](https://forums.developer.nvidia.com/t/nvml-12-535-43-02-breaks-backwards-compatibility/254999/17). In some broken versions v3 used a different data structure (`nvmlProcessInfo_v2_st` with *different* byte size), and suddenly removed & reverted that, etc. After everything is resolved (535.104.05+), the v3 API uses the same struct `nvmlProcessInfo_v2_st` as v2 API so I think they are basically the same.

Looping @XuehaiPan here, please feel free to correct if I said anything wrong.

Comment (XuehaiPan) [reply dashboard/modules/reporter/reporter_agent.py:418]: Agree with @wookayin. To support a wide range of NVIDIA driver verions without complex monkey patching, we'd better use `nvmlDeviceGet{Compute,Graphics}RunningProcesses_v2` APIs with `nvmlProcessInfo_v2` type.

File: dashboard/modules/reporter/reporter_agent.py
Comment (XuehaiPan) [inline dashboard/modules/reporter/reporter_agent.py:444]: We could merge these two `nvmlDeviceGetMemoryInfo` calls into one.

Comment (XuehaiPan) [reply dashboard/modules/reporter/reporter_agent.py:444]: FYI, in NVIDIA driver 510.39.01, a v2 memory info API was added:

https://github.com/NVIDIA/nvidia-settings/blob/510.39.01/src/nvml.h#L218-L241

The unversioned API (v1) and v2 API will return different results on R510+ drivers.

File: python/ray/_private/thirdparty/pynvml/__init__.py
Comment (jjyao) [inline python/ray/_private/thirdparty/pynvml/__init__.py:2]: Add a comment saying why we pick this version: something like we are using this version because it uses v2 api and supports a wider range of drivers.

**Review by jjyao — APPROVED** [review]

**Review by ericl — APPROVED** [review]
Always great to see deps removed

**Pull request merged**
